include/picongpu/plugins/openPMD/openPMDWriter.hpp

```diff
@@ -0,0 +1,1446 @@
+                    sizeof( ComponentType ),
+                    ::openPMD::determineDatatype< ComponentType >(),
+                    components,
+                    getName(),
+                    fieldTmp->getHostDataBox().getPointer(),
+                    getUnit(),
+                    FieldTmp::getUnitDimension< Solver >(),
+                    std::move( inCellPosition ),
+                    timeOffset,
+                    isDomainBound );
+
+                dc.releaseData( FieldTmp::getUniqueId( 0 ) );
+            }
+        };
+
+    public:
+        /** constructor
+         *
+         * @param help instance of the class Help
+         * @param id index of this plugin instance within help
+         * @param cellDescription PIConGPu cell description information for
+         * kernel index mapping
+         */
+        openPMDWriter(
+            std::shared_ptr< plugins::multi::IHelp > & help,
+            size_t const id,
+            MappingDesc * cellDescription ) :
+            m_help( std::static_pointer_cast< Help >( help ) ),
+            m_id( id ),
+            m_cellDescription( cellDescription ),
+            outputDirectory( "openPMD" ),
+            lastSpeciesSyncStep( pmacc::traits::limits::Max< uint32_t >::value )
+        {
+            mThreadParams.compressionMethod = m_help->compression.get( id );
+
+            GridController< simDim > & gc =
+                Environment< simDim >::get().GridController();
+            /* It is important that we never change the mpi_pos after this point
+             * because we get problems with the restart.
+             * Otherwise we do not know which gpu must load the ghost parts
+             * around the sliding window.
+             */
+            mpi_pos = gc.getPosition();
+            mpi_size = gc.getGpuNodes();
+
+            if( m_help->selfRegister )
+            {
+                std::string notifyPeriod = m_help->notifyPeriod.get( id );
+                /* only register for notify callback when .period is set on
+                 * command line */
+                if( !notifyPeriod.empty() )
+                {
+                    Environment<>::get()
+                        .PluginConnector()
+                        .setNotificationPeriod( this, notifyPeriod );
+
+                    /** create notify directory */
+                    Environment< simDim >::get()
+                        .Filesystem()
+                        .createDirectoryWithPermissions( outputDirectory );
+                }
+            }
+
+            // avoid deadlock between not finished pmacc tasks and mpi blocking
+            // collectives
+            __getTransactionEvent().waitForFinished();
+            mThreadParams.communicator = MPI_COMM_NULL;
+            MPI_CHECK( MPI_Comm_dup(
+                gc.getCommunicator().getMPIComm(),
+                &( mThreadParams.communicator ) ) );
+
+            /* TODO: select MPI method, #OSTs and #aggregators */
```

Comment:
Please use `@todo` if this todo is still valid or remove this comment.

Is this not handled in OpenPMD directly?

Reply:
Yep, I removed it.